includes/data-events/class-data-events.php

```diff
@@ -0,0 +1,230 @@
+<?php
+/**
+ * Newspack Data Events.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack;
+
+use WP_Error;
+
+/**
+ * Main class.
+ */
+final class Data_Events {
+	/**
+	 * Asynchronous action name.
+	 */
+	const ACTION = 'newspack_data_event';
+
+	/**
+	 * Registered callable handlers, keyed by their action name.
+	 *
+	 * @var callable[]
+	 */
+	private static $actions = [];
+
+	/**
+	 * Registered global callable handlers to be executed on all actions.
+	 *
+	 * @var callable[]
+	 */
+	private static $global_handlers = [];
+
+	/**
+	 * Initialize hooks.
+	 */
+	public static function init() {
+		\add_action( 'wp_ajax_' . self::ACTION, [ __CLASS__, 'maybe_handle' ] );
+		\add_action( 'wp_ajax_nopriv_' . self::ACTION, [ __CLASS__, 'maybe_handle' ] );
+	}
+
+	/**
+	 * Maybe handle an event.
+	 */
+	public static function maybe_handle() {
+		// Don't lock up other requests while processing.
+		session_write_close(); // phpcs:ignore
+
+		if ( ! isset( $_REQUEST['nonce'] ) || ! \wp_verify_nonce( \sanitize_text_field( $_REQUEST['nonce'] ), self::ACTION ) ) {
+			wp_die();
+		}
+
+		$action_name = isset( $_POST['action_name'] ) ? \sanitize_text_field( \wp_unslash( $_POST['action_name'] ) ) : null;
+		if ( ! $action_name || ! isset( self::$actions[ $action_name ] ) ) {
+			wp_die();
+		}
+
+		$timestamp = isset( $_POST['timestamp'] ) ? \sanitize_text_field( \wp_unslash( $_POST['timestamp'] ) ) : null;
+		$data      = isset( $_POST['data'] ) ? $_POST['data'] : null; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$client_id = isset( $_POST['client_id'] ) ? \sanitize_text_field( \wp_unslash( $_POST['client_id'] ) ) : null;
+
+		self::handle( $action_name, $timestamp, $data, $client_id );
+
+		wp_die();
+	}
+
+	/**
+	 * Handle an event.
+	 *
+	 * @param string $action_name Action name.
+	 * @param int    $timestamp   Timestamp.
+	 * @param mixed  $data        Data.
+	 * @param string $client_id   Client ID.
+	 */
+	public static function handle( $action_name, $timestamp, $data, $client_id ) {
+		// Execute global handlers.
+		Logger::log( 'Executing global action handlers.' );
+		foreach ( self::$global_handlers as $handler ) {
+			try {
+				call_user_func( $handler, $action_name, $timestamp, $data, $client_id );
+			} catch ( \Throwable $e ) {
+				// Catch fatal errors so it doesn't disrupt other handlers.
+				Logger::error( $e->getMessage() );
+			}
+		}
+
+		// Execute action handlers.
+		Logger::log( 'Executing action handler: ' . $action_name );
+		$handlers = self::get_action_handlers( $action_name );
+		foreach ( $handlers as $handler ) {
+			try {
+				call_user_func( $handler, $timestamp, $data, $client_id );
+			} catch ( \Throwable $e ) {
+				// Catch fatal errors so it doesn't disrupt other handlers.
+				Logger::error( $e->getMessage() );
+			}
+		}
+
+		// Execute WP hooks.
+		do_action( self::ACTION, $action_name, $timestamp, $data, $client_id );
```

Comment:
We should add a doc block for each action.

Also, I wouldn't make the action name dynamic like that. If we ever want to change the class constant for any reason this would break any consumer of this library using the hook previous names...

Also, it makes it slightly harder for a programmer to find where this action is being fired from (you could workaround it by adding the name to the comment to make it discoverable...) But I think in general it's not a good thing to make the whole action name dynamic.

Reply:
Good call! 61ae87cc7de97b807a3547dbf204270b00d1c95d improves readability and docs for all WP hooks.